jctools-core/src/main/java/org/jctools/queues/SpscUnboundedArrayQueue.java

```diff
@@ -15,72 +15,19 @@
 
 import static org.jctools.util.UnsafeAccess.UNSAFE;
 
-import java.lang.reflect.Field;
-import java.util.AbstractQueue;
-import java.util.Iterator;
-
 import org.jctools.util.Pow2;
 import org.jctools.util.UnsafeAccess;
+import org.jctools.util.UnsafeRefArrayAccess;
 
-abstract class SpscUnboundedArrayQueuePrePad<E> extends AbstractQueue<E> {
-    long p0, p1, p2, p3, p4, p5, p6, p7;
-    long p10, p11, p12;
-    // p13, p14, p15, p16, p17; drop 4 longs, the cold fields act as buffer
-}
-abstract class SpscUnboundedArrayQueueProducerColdFields<E> extends SpscUnboundedArrayQueuePrePad<E> {
-    protected int producerLookAheadStep;
-    protected long producerLookAhead;
-    protected long producerMask;
-    protected E[] producerBuffer;
-}
-abstract class SpscUnboundedArrayQueueProducerFields<E> extends SpscUnboundedArrayQueueProducerColdFields<E> {
-    protected long producerIndex;
-}
-
-abstract class SpscUnboundedArrayQueueL2Pad<E> extends SpscUnboundedArrayQueueProducerFields<E> {
-    long p0, p1, p2, p3, p4, p5, p6, p7;
-    long p10, p11, p12, p13, p14, p15, p16, p17;
-}
-abstract class SpscUnboundedArrayQueueConsumerColdField<E> extends SpscUnboundedArrayQueueL2Pad<E> {
-    protected long consumerMask;
-    protected E[] consumerBuffer;
-}
-
-abstract class SpscUnboundedArrayQueueConsumerField<E> extends SpscUnboundedArrayQueueConsumerColdField<E> {
-    protected long consumerIndex;
-}
-
-public class SpscUnboundedArrayQueue<E> extends SpscUnboundedArrayQueueConsumerField<E>
-    implements QueueProgressIndicators{
-    static final int MAX_LOOK_AHEAD_STEP = Integer.getInteger("jctools.spsc.max.lookahead.step", 4096);
-    private final static long P_INDEX_OFFSET;
-    private final static long C_INDEX_OFFSET;
+public class SpscUnboundedArrayQueue<E> extends AbstractSpscResizingArrayQueue<E>
+    {
+    private static final int MAX_LOOK_AHEAD_STEP = Integer.getInteger("jctools.spsc.max.lookahead.step", 4096);
     private static final long REF_ARRAY_BASE;
-    private static final int REF_ELEMENT_SHIFT;
+    private static final int REF_ELEMENT_SHIFT = UnsafeRefArrayAccess.REF_ELEMENT_SHIFT;
```

Comment:
Is this at all a concern? The static block initialising it is a copy-paste duplicate of the one in UnsafeRefArrayAccess so it seemed appropriate to just reuse the value computed there (since this class is coupled to UnsafeRefArrayAccess anyway).